### PR TITLE
Eagerly search all segments for an index block

### PR DESCRIPTION
**Pull request by ryanhall07**
This is part of a larger refactoring to chunk up the processing of a single
block. Each chunk would need to acquire an index worker from the pool.
This prevents large queries with many documents results in an index
block from starving smaller queries. Ideally every chunk of work would
hold an index worker for the same amount of time.

In this change we eagerly search all the segments for a block on the
first call to Next(). Metrics show the time search is relatively
constant across queries, so it shouldn't add much overhead to the first
Next() call. Once the postings lists are eagerly retrived, the doc
iterator can iterate over them in batches and yield the worker after N
docs. Once it acquires a worker again it can resume the iteration.

**What this PR does / why we need it**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

**Review by nbroyles — APPROVED** [review]

File: src/m3ninx/search/executor/iterator.go [hunk 1]
```diff
@@ -28,77 +28,73 @@ import (
...
+// doc iterators. for each posting in the list, the encoded document is retrieved.
 type iterator struct {
-	ctx      context.Context
-	searcher search.Searcher
-	readers  index.Readers
-
-	idx                 int
-	currDoc             doc.Document
-	currIter            doc.Iterator
+	// immutable state
```
Comment (robskillington) [inline src/m3ninx/search/executor/iterator.go:44]: +1 for describing mutable vs immutable.

**Review by robskillington — APPROVED** [review]
LGTM

**Comment by robskillington** [comment]
What do we think about this pattern of passing in ctx just for sake of creating a trace span btw? Do you think we should pass in a trimmed interface which can generate the spans? Just feels a little strange passing ctx here, I feel there's some expectation that the datastructure should check the ctx isn't done, etc. But yeah, shrug.

**Comment by ryanhall07** [comment]
> What do we think about this pattern of passing in ctx just for sake of creating a trace span btw? Do you think we should pass in a trimmed interface which can generate the spans? Just feels a little strange passing ctx here, I feel there's some expectation that the datastructure should check the ctx isn't done, etc. But yeah, shrug.

the problem is opentracing is going to want a std ctx.

**Pull request merged**